### PR TITLE
[FLINK-25943][connector/common] Add buffered requests to snapshot state in AsyncSyncWriter.

### PR DESCRIPTION
**Pull request by vahmed-hamdy**
## What is the purpose of the change
- Adding buffered requests in snapshot state of `AsyncSinkWriter` and its implementation so that the writers can resume the buffered requests when restoring from a checkpoint/savepoint.

## Brief change log

  - Added `BufferedRequestsState` in `flink-connector-base`.
  - Added `AsyncSinkWriterStateSerializer` in `flink-connector-base`.
  - implemented `AsyncSinkWriterStateSerializer` in both `flink-connector-aws-kinesis-data-streams` and `flink-connector-aws-kinesis-firehose`.
  
## Verifying this change

- Added unit tests for new serializer class.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by vahmed-hamdy** [comment]
@flinkbot run azure

**Comment by vahmed-hamdy** [comment]
@flinkbot run azure

**Comment by vahmed-hamdy** [comment]
@flinkbot run azure

**Comment by vahmed-hamdy** [comment]
@flinkbot run azure

**Comment by vahmed-hamdy** [comment]
@flinkbot run azure

**Comment by CrynetLogistics** [comment]
@flinkbot run azure

**Comment by CrynetLogistics** [comment]
@flinkbot run azure

**Comment by CrynetLogistics** [comment]
@flinkbot run azure

**Comment by CrynetLogistics** [comment]
@flinkbot run azure

**Comment by CrynetLogistics** [comment]
@flinkbot run azure

**Comment by CrynetLogistics** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:46]: We are not setting the partition key here. Also, this implementation seems like it needs to be tied to the `ElementConverter`, somehow

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:38]: Instead of just ripping out the data, can we serialise the whole object instead?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:419]: How about other validations such as number of buffered records (`maxBufferedRequests`)? Maybe we should allow this condition

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:429]: Is this the right time to flush? Would this state restore be called during operator `initializing` state? If so, we probably do not want to flush here

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:425]: nit: `sum += wrapper.getSize();`

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:34]: Why is this `@PublicEvolving`?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:30]: Please improve this comment

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:37]: This can be `long`

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:79]: Why did you implement this? Flink coding standard advise against this unless really necessary https://flink.apache.org/contributing/code-style-and-quality-java.html

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/RequestEntryWrapper.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/RequestEntryWrapper.java:70]: Same question as above, why add this?

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSink.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSink.java:135]: Yes, we wrap the entire state in a single item.
I agree, will validate and fail

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java:46]: Use a multiline javadoc here instead of multiple single line comments

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:425]: Can you use `state.getStateSize()` instead?

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:34]: I wanted to use `@Internal` but this fails architecture tests.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:419]: We allowed `maxBufferedRequests` to flush at the end of initialisation.
It is agreed not to validate and to drop the flush logic.

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:35]: It was intended for readability, we would other wise be transfering state as `Collection<RequestEntryWrapper<RequestEntryT>>>` , I am fine removing it.

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:60]: These methods are duplicated/could be combined.

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:88]: These methods are duplicated/could be combined.

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:42]: Agree, It is specific to serialize `BufferRequestsState` whose request would never have an explicit hash key.
will remove and assert.

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:52]: nit: I think we can be more specific here. Can we update to suggest that the `ElementConverter` has evolved without updating the state serialiser?

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:71]: Since we have helper methods in parent `readInt`/`readLong` we could make this `readString` ?

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:46]: We should specify the charset so data is portable between systems

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java:438]: Can we validate that there are no more than 1? Do we already do this somewhere?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java:99]: If we are going to rethrow we should wrap `e`. However, I would be inclined to just remove these methods `readLong`/`readInt` directly and let the exceptions bubble up. This is similar to other implementations. Generally speaking it is bad practise to catch `RuntimeException`

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:66]: This should be `long` not `Long`

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializerTest.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializerTest.java:55]: This is duplicated in a few places. Can we move to a test util? Also can simplify with java streams as pointed out in a nit comment above

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:71]: It is agreed to remove byte stream validation and bubble up the original error.

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:75]: We need to confirm that this will always read the entire length of the string.

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/test/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializerTest.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-kinesis-data-streams/src/test/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializerTest.java:74]: nit: Add new lines to make this more readable

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterStateSerializer.java:80]: nit: Add new lines to make more readable

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:44]: nit: 
```
this.bufferedRequestEntries = new ArrayList<>(bufferedRequestEntries);
```

File: flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java
Comment (dannycranmer) [reply flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsStateSerializer.java:75]: I see this is a wrapped byte array in the parent class

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:59]: Please see guidance on use of java streams. Unfortunately I would consider this performance critical and therefore please remove streams: https://flink.apache.org/contributing/code-style-and-quality-java.html#java-streams

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:37]: Why is this failing?

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTestUtils.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriterTestUtils.java:59]: nit: add whitespace to make this more readable

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:37]: it voilated `PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES` arch rule
the `snapshot()` function in the writer is publich and can't return an `@Internal` class

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (CrynetLogistics) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:59]: Interesting... noted.

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (dannycranmer) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:50]: Could be combined with `Collection` interface

**Review by dannycranmer — APPROVED** [review]
LGTM, will merge once we have green light from CI

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java
Comment (vahmed-hamdy) [reply flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/BufferedRequestState.java:50]: They were already, were split out to ensure an ordering data structure.

**Pull request merged**
